File: backend/app/models/reconciliation.py
"""
Bank file and journal voucher reconciliation.

This is the control that closes payroll. Validation asks whether the register
is *correct*; reconciliation asks whether what was correct is what actually
left the bank and what actually reached the ledger. Those are three different
numbers in most companies, and the gaps between them are where both errors and
fraud live:

* the register says ₹868 lakh of net pay is due;
* the bank file says ₹868.4 lakh was paid, to 1,204 accounts;
* the journal voucher posted ₹9.1 crore of cost to the general ledger.

Nothing in a payroll system reconciles those for you, because the payroll
system produced all three and has no independent view. This product never runs
payroll, which is exactly why it can hold the three against each other.

Why so much configuration
-------------------------
Two things here are different at every single company, and a module that
assumes either one is useless at the second client:

**Bank files.** Every bank's bulk-payment layout differs, and so does the
subset of columns a given corporate account is set up to send. Account number
in column 3 at one bank and column 7 at another; amounts in rupees here and
paise there; a header and two trailer rows at one bank and neither at the next.
So a bank format is *data* — :class:`BankFileProfile` — not code.

**Journal vouchers.** There is no standard payroll JV. One company posts a
single consolidated voucher; another posts one per cost centre. One debits
gratuity as an expense monthly; another only on payment. Chart-of-accounts
codes are unique to the company by definition. One exports to Tally, the next
to SAP. So a JV format is also data — a :class:`JvTemplate` of ordered
:class:`JvRule` lines, each mapping payroll cost measures onto an account and
a side.

Because a JV rule names measures from ``services/cost_model``, the voucher and
the cost dashboard are built from one set of numbers. They cannot drift apart:
if the dashboard says payroll cost ₹9.1 crore, a balanced JV built from the
same measures posts ₹9.1 crore.
"""
import uuid
from datetime import date, datetime
from decimal import Decimal
from typing import Any

from sqlalchemy import (
    JSON,
    Boolean,
    Date,
    DateTime,
    ForeignKey,
    Integer,
    Numeric,
    String,
    Text,
    UniqueConstraint,
    Uuid,
    func,
)
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database import Base


# ---------------------------------------------------------------------------
# Bank side: how to read this company's bank file
# ---------------------------------------------------------------------------
class BankFileProfile(Base):
    """
    A named, editable description of one bank file layout.

    Held per entity rather than globally: a bureau's clients bank with
    different banks, and two clients at the same bank can be provisioned with
    different column sets.
    """

    __tablename__ = "bank_file_profiles"
    __table_args__ = (UniqueConstraint("entity_id", "name"),)

    id: Mapped[uuid.UUID] = mapped_column(Uuid, primary_key=True, default=uuid.uuid4)
    entity_id: Mapped[uuid.UUID] = mapped_column(
        Uuid, ForeignKey("entities.id", ondelete="CASCADE"), nullable=False, index=True
    )
    name: Mapped[str] = mapped_column(String(120), nullable=False)
    bank_label: Mapped[str | None] = mapped_column(String(120))
    note: Mapped[str | None] = mapped_column(Text)

    # What kind of file this reads. A payment advice is what goes *to* the bank;
    # a statement or a return file is what comes *back*. They reconcile against
    # different things, so the kind is part of the profile rather than guessed.
    kind: Mapped[str] = mapped_column(String(32), nullable=False, default="payment_advice")

    # ---- layout ---------------------------------------------------------
    layout: Mapped[str] = mapped_column(String(16), nullable=False, default="delimited")
    delimiter: Mapped[str] = mapped_column(String(4), nullable=False, default=",")
    encoding: Mapped[str] = mapped_column(String(32), nullable=False, default="utf-8")
    has_header: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)
    skip_rows: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    # Banks routinely end a file with a control total. Counting it as a payment
    # inflates the file total by roughly double, so it is dropped by count.
    trailer_rows: Mapped[int] = mapped_column(Integer, nullable=False, default=0)

    # {field: column name, 1-based index, or [start, length] for fixed width}
    column_map: Mapped[dict[str, Any]] = mapped_column(JSON, nullable=False, default=dict)

    # ---- value handling -------------------------------------------------
    amount_unit: Mapped[str] = mapped_column(String(16), nullable=False, default="rupees")
    amount_sign: Mapped[str] = mapped_column(String(24), nullable=False, default="as_is")
    date_format: Mapped[str | None] = mapped_column(String(32))
    # Employee codes are the join key, and they are rarely clean on both sides.
    # "EMP0042" in the register and "42" in the bank file is the same person.
    employee_id_transform: Mapped[str] = mapped_column(String(32), nullable=False, default="trim")

    # Keep only rows whose column matches: {"column": "txn_type", "in": ["SAL"]}
    row_filter: Mapped[dict[str, Any]] = mapped_column(JSON, nullable=False, default=dict)

    is_default: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    created_by_user_id: Mapped[uuid.UUID | None] = mapped_column(
        Uuid, ForeignKey("users.id", ondelete="SET NULL")
    )
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
    updated_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), onupdate=func.now())


class BankFile(Base):
    """One uploaded bank file, kept so a reconciliation stays reproducible."""

    __tablename__ = "bank_files"

    id: Mapped[uuid.UUID] = mapped_column(Uuid, primary_key=True, default=uuid.uuid4)
    entity_id: Mapped[uuid.UUID] = mapped_column(
        Uuid, ForeignKey("entities.id", ondelete="CASCADE"), nullable=False, index=True
    )
    profile_id: Mapped[uuid.UUID | None] = mapped_column(
        Uuid, ForeignKey("bank_file_profiles.id", ondelete="SET NULL")
    )
    period_month: Mapped[date] = mapped_column(Date, nullable=False, index=True)
    kind: Mapped[str] = mapped_column(String(32), nullable=False, default="payment_advice")
    filename: Mapped[str | None] = mapped_column(String(512))
    row_count: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    total_amount: Mapped[Decimal] = mapped_column(Numeric(18, 2), nullable=False, default=Decimal("0"))
    # What the file's own trailer claimed, where it had one. A file whose stated
    # total disagrees with the sum of its own rows is truncated or edited, and
    # that is worth knowing before any of it is reconciled.
    stated_total: Mapped[Decimal | None] = mapped_column(Numeric(18, 2))
    problems: Mapped[list[Any]] = mapped_column(JSON, nullable=False, default=list)
    note: Mapped[str | None] = mapped_column(Text)

    uploaded_by_user_id: Mapped[uuid.UUID | None] = mapped_column(
        Uuid, ForeignKey("users.id", ondelete="SET NULL")
    )
    uploaded_by_email: Mapped[str | None] = mapped_column(String(255))
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())

    rows = relationship("BankFileRow", back_populates="file", cascade="all, delete-orphan")


class BankFileRow(Base):
    """One payment line as the bank file stated it."""

    __tablename__ = "bank_file_rows"

    id: Mapped[uuid.UUID] = mapped_column(Uuid, primary_key=True, default=uuid.uuid4)
    file_id: Mapped[uuid.UUID] = mapped_column(
        Uuid, ForeignKey("bank_files.id", ondelete="CASCADE"), nullable=False, index=True
    )
    entity_id: Mapped[uuid.UUID] = mapped_column(
        Uuid, ForeignKey("entities.id", ondelete="CASCADE"), nullable=False, index=True
    )
    period_month: Mapped[date] = mapped_column(Date, nullable=False, index=True)
    row_number: Mapped[int] = mapped_column(Integer, nullable=False, default=0)

    employee_id: Mapped[str | None] = mapped_column(String(64), index=True)
    employee_name: Mapped[str | None] = mapped_column(String(255))
    account_number: Mapped[str | None] = mapped_column(String(64))
    ifsc: Mapped[str | None] = mapped_column(String(16))
    amount: Mapped[Decimal] = mapped_column(Numeric(16, 2), nullable=False, default=Decimal("0"))
    reference: Mapped[str | None] = mapped_column(String(128))
    status: Mapped[str | None] = mapped_column(String(64))
    value_date: Mapped[date | None] = mapped_column(Date)
    # The line exactly as read, so an operator can see what the profile did to
    # their file when a mapping turns out to be wrong.
    raw: Mapped[dict[str, Any]] = mapped_column(JSON, nullable=False, default=dict)

    file = relationship("BankFile", back_populates="rows")


# ---------------------------------------------------------------------------
# Ledger side: how this company builds its journal voucher
# ---------------------------------------------------------------------------
class JvTemplate(Base):
    """
    One company's way of turning a month of payroll cost into a voucher.

    Approved rather than merely saved, for the same reason a budget is: the
    voucher that goes to the ledger is a number the auditor will ask about, and
    "someone edited the mapping in March" has to be answerable.
    """

    __tablename__ = "jv_templates"
    __table_args__ = (UniqueConstraint("entity_id", "name"),)

    id: Mapped[uuid.UUID] = mapped_column(Uuid, primary_key=True, default=uuid.uuid4)
    entity_id: Mapped[uuid.UUID] = mapped_column(
        Uuid, ForeignKey("entities.id", ondelete="CASCADE"), nullable=False, index=True
    )
    name: Mapped[str] = mapped_column(String(120), nullable=False)
    note: Mapped[str | None] = mapped_column(Text)
    state: Mapped[str] = mapped_column(String(16), nullable=False, default="draft")

    # ---- the options that differ company to company ---------------------
    # accrual: gratuity and every employer contribution hit the month they are
    #   earned, which is what Ind AS expects.
    # cash: only what is actually disbursed or remitted. Smaller companies post
    #   this way and provide for gratuity annually instead.
    posting_basis: Mapped[str] = mapped_column(String(16), nullable=False, default="accrual")
    # consolidated: one voucher, each line optionally tagged with a cost centre.
    # per_group: a separate voucher per cost centre, which is what companies
    #   whose ERP posts by profit centre need.
    split_mode: Mapped[str] = mapped_column(String(16), nullable=False, default="consolidated")
    # Which dimension is the cost centre here. Companies disagree: some post by
    # department, some by the cost_center code, some by location.
    group_by: Mapped[str | None] = mapped_column(String(32))
    # summary: one line per account. by_component: one per measure within the
    # account. by_employee: employee-wise, which a few small companies still do.
    detail_level: Mapped[str] = mapped_column(String(16), nullable=False, default="summary")
    # two_column: separate debit and credit columns. signed: one amount column
    # with credits negative, which most ERP imports want.
    sign_convention: Mapped[str] = mapped_column(String(16), nullable=False, default="two_column")
    # Which net-pay figure the salary-payable credit is built from. "computed"
    # is gross less the deductions in the taxonomy, and always balances against
    # the cost side. "stated" is the net the register itself declared, which is
    # what the bank was actually told to pay — truer to the disbursement, but it
    # will not balance where the register deducts something it does not itemise,
    # such as a salary advance. Companies genuinely do both, so it is a choice
    # rather than a default, and the difference is always reported either way.
    net_pay_source: Mapped[str] = mapped_column(String(16), nullable=False, default="computed")
    voucher_date_rule: Mapped[str] = mapped_column(String(24), nullable=False, default="month_end")
    voucher_type: Mapped[str] = mapped_column(String(32), nullable=False, default="Journal")
    narration_template: Mapped[str] = mapped_column(
        String(255), nullable=False, default="Payroll for {period_label}{scope_suffix}"
    )
    export_format: Mapped[str] = mapped_column(String(24), nullable=False, default="generic_csv")

    # A rupee or two of rounding across thousands of employees is ordinary; a
    # lakh is a mapping error. The tolerance separates the two, and what happens
    # to a within-tolerance difference is itself a choice.
    balance_tolerance: Mapped[Decimal] = mapped_column(
        Numeric(12, 2), nullable=False, default=Decimal("1.00")
    )
    rounding_mode: Mapped[str] = mapped_column(String(16), nullable=False, default="none")
    rounding_account: Mapped[str | None] = mapped_column(String(64))

    is_current: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    created_by_user_id: Mapped[uuid.UUID | None] = mapped_column(
        Uuid, ForeignKey("users.id", ondelete="SET NULL")
    )
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
    approved_by_user_id: Mapped[uuid.UUID | None] = mapped_column(
        Uuid, ForeignKey("users.id", ondelete="SET NULL")
    )
    approved_by_email: Mapped[str | None] = mapped_column(String(255))
    approved_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))

    rules = relationship(
        "JvRule",
        back_populates="template",
        cascade="all, delete-orphan",
        order_by="JvRule.sequence",
    )


class JvRule(Base):
    """
    One posting line: these measures, to this account, on this side.

    ``measures`` names keys from ``services.cost_model`` — the same taxonomy the
    cost dashboard reports. That is the whole point of the indirection: an
    account mapped to ``er_pf`` posts exactly the employer PF the dashboard
    shows, so the ledger and the dashboard cannot tell two different stories.
    """

    __tablename__ = "jv_rules"

    id: Mapped[uuid.UUID] = mapped_column(Uuid, primary_key=True, default=uuid.uuid4)
    template_id: Mapped[uuid.UUID] = mapped_column(
        Uuid, ForeignKey("jv_templates.id", ondelete="CASCADE"), nullable=False, index=True
    )
    entity_id: Mapped[uuid.UUID] = mapped_column(
        Uuid, ForeignKey("entities.id", ondelete="CASCADE"), nullable=False, index=True
    )
    sequence: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    label: Mapped[str] = mapped_column(String(160), nullable=False)
    account_code: Mapped[str] = mapped_column(String(64), nullable=False, default="")
    account_name: Mapped[str | None] = mapped_column(String(160))
    side: Mapped[str] = mapped_column(String(8), nullable=False, default="debit")
    measures: Mapped[list[Any]] = mapped_column(JSON, nullable=False, default=list)
    # Restrict this line to part of the workforce: {"department": ["Sales"]}.
    # A company that books field staff to a different account needs this; one
    # that does not leaves it empty.
    filters: Mapped[dict[str, Any]] = mapped_column(JSON, nullable=False, default=dict)
    # Tag each posted line with a dimension value, overriding the template's
    # own grouping for this line only.
    cost_center_from: Mapped[str | None] = mapped_column(String(32))
    active: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)
    note: Mapped[str | None] = mapped_column(Text)

    template = relationship("JvTemplate", back_populates="rules")


# ---------------------------------------------------------------------------
# The reconciliation itself
# ---------------------------------------------------------------------------
class ReconRun(Base):
    """
    A reconciliation someone ran and kept.

    Reconciliation is only a control if it leaves a record. A run stores what
    was compared, what came out, and who signed it off, so "payroll was
    reconciled for June" is a fact with evidence behind it rather than a claim.
    """

    __tablename__ = "recon_runs"

    id: Mapped[uuid.UUID] = mapped_column(Uuid, primary_key=True, default=uuid.uuid4)
    entity_id: Mapped[uuid.UUID] = mapped_column(
        Uuid, ForeignKey("entities.id", ondelete="CASCADE"), nullable=False, index=True
    )
    period_month: Mapped[date] = mapped_column(Date, nullable=False, index=True)
    # "bank" | "jv" | "both"
    kind: Mapped[str] = mapped_column(String(16), nullable=False, default="bank")
    bank_file_id: Mapped[uuid.UUID | None] = mapped_column(
        Uuid, ForeignKey("bank_files.id", ondelete="SET NULL")
    )
    jv_template_id: Mapped[uuid.UUID | None] = mapped_column(
        Uuid, ForeignKey("jv_templates.id", ondelete="SET NULL")
    )
    state: Mapped[str] = mapped_column(String(16), nullable=False, default="open")
    summary: Mapped[dict[str, Any]] = mapped_column(JSON, nullable=False, default=dict)

    created_by_user_id: Mapped[uuid.UUID | None] = mapped_column(
        Uuid, ForeignKey("users.id", ondelete="SET NULL")
    )
    created_by_email: Mapped[str | None] = mapped_column(String(255))
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
    closed_by_email: Mapped[str | None] = mapped_column(String(255))
    closed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))

    exceptions = relationship(
        "ReconException", back_populates="run", cascade="all, delete-orphan"
    )


class ReconException(Base):
    """One thing that did not reconcile, and by how much."""

    __tablename__ = "recon_exceptions"

    id: Mapped[uuid.UUID] = mapped_column(Uuid, primary_key=True, default=uuid.uuid4)
    run_id: Mapped[uuid.UUID] = mapped_column(
        Uuid, ForeignKey("recon_runs.id", ondelete="CASCADE"), nullable=False, index=True
    )
    entity_id: Mapped[uuid.UUID] = mapped_column(
        Uuid, ForeignKey("entities.id", ondelete="CASCADE"), nullable=False, index=True
    )
    code: Mapped[str] = mapped_column(String(64), nullable=False, index=True)
    severity: Mapped[str] = mapped_column(String(16), nullable=False, default="medium")
    title: Mapped[str] = mapped_column(String(255), nullable=False)
    detail: Mapped[str | None] = mapped_column(Text)
    employee_id: Mapped[str | None] = mapped_column(String(64), index=True)
    employee_name: Mapped[str | None] = mapped_column(String(255))
    scope: Mapped[str | None] = mapped_column(String(255))
    expected: Mapped[Decimal | None] = mapped_column(Numeric(16, 2))
    actual: Mapped[Decimal | None] = mapped_column(Numeric(16, 2))
    difference: Mapped[Decimal | None] = mapped_column(Numeric(16, 2))
    context: Mapped[dict[str, Any]] = mapped_column(JSON, nullable=False, default=dict)

    run = relationship("ReconRun", back_populates="exceptions")

File: frontend/src/lib/reconciliation.ts
import { apiFetch, parseEnvelopeResponse } from "@/lib/api";

/**
 * Bank file and journal voucher reconciliation.
 *
 * Every shape here mirrors the API exactly, including the parts that report an
 * *absence*: a month with no bank file comes back with `ready: false` and a
 * sentence saying so, and the UI shows that sentence rather than an empty
 * panel. An empty panel reads as a clean month.
 */

// ---------------------------------------------------------------------------
// Shared
// ---------------------------------------------------------------------------
export type Severity = "high" | "medium" | "low";

export type ReconException = {
  code: string;
  label: string;
  severity: Severity;
  title: string;
  detail: string | null;
  meaning: string | null;
  action: string | null;
  employee_id: string | null;
  employee_name: string | null;
  scope: string | null;
  expected: number | null;
  actual: number | null;
  difference: number | null;
  context: Record<string, unknown>;
};

export type ExceptionCounts = {
  total: number;
  by_severity: Record<Severity, number>;
  by_code: Record<string, number>;
};

export type ExceptionKind = {
  code: string;
  label: string;
  severity: Severity;
  meaning: string;
  action: string;
};

// ---------------------------------------------------------------------------
// Bank side
// ---------------------------------------------------------------------------
export type BankFieldMeta = {
  key: string;
  label: string;
  required: boolean;
  hint: string;
};

export type OptionMeta = { key: string; label: string; hint?: string };

export type BankFieldCatalogue = {
  fields: BankFieldMeta[];
  kinds: OptionMeta[];
  layouts: OptionMeta[];
  amount_units: OptionMeta[];
  amount_signs: OptionMeta[];
  id_transforms: OptionMeta[];
};

export type BankProfile = {
  id: string;
  name: string;
  bank_label: string | null;
  note: string | null;
  kind: string;
  layout: string;
  delimiter: string;
  encoding: string;
  has_header: boolean;
  skip_rows: number;
  trailer_rows: number;
  column_map: Record<string, string | number | number[]>;
  amount_unit: string;
  amount_sign: string;
  date_format: string | null;
  employee_id_transform: string;
  row_filter: Record<string, unknown>;
  is_default: boolean;
  created_at: string | null;
};

export type BankPreset = Omit<BankProfile, "id" | "is_default" | "created_at"> & {
  key: string;
};

export type BankFileMeta = {
  id: string;
  period: string;
  period_label: string;
  kind: string;
  filename: string | null;
  row_count: number;
  total: number;
  stated_total: number | null;
  problems: string[];
  note: string | null;
  profile_id: string | null;
  uploaded_by: string | null;
  created_at: string | null;
};

export type BankFileRow = {
  row_number: number;
  employee_id: string | null;
  employee_name: string | null;
  account_number: string | null;
  ifsc: string | null;
  amount: number;
  reference: string | null;
  status: string | null;
  value_date: string | null;
};

export type ProfileTest = {
  filename: string;
  row_count: number;
  skipped_by_filter: number;
  total: number;
  stated_total: number | null;
  header: string[];
  problems: string[];
  rows: BankFileRow[];
  truncated: boolean;
};

export type MappingSuggestion = {
  header: string[];
  sample: string[][];
  column_map: Record<string, string>;
  unmapped_columns?: string[];
  note: string;
};

export type BankReconciliation = {
  period: string;
  period_label: string;
  file: {
    id: string;
    filename: string | null;
    kind: string;
    row_count: number;
    total: number;
    stated_total: number | null;
  };
  summary: {
    register_employees: number;
    bank_rows: number;
    matched: number;
    unmatched_in_register: number;
    unmatched_in_file: number;
    due_total: number;
    paid_total: number;
    matched_total: number;
    difference: number;
    reconciled: boolean;
    tolerance: number;
  };
  counts: ExceptionCounts;
  exceptions: ReconException[];
};

// ---------------------------------------------------------------------------
// Ledger side
// ---------------------------------------------------------------------------
export type JvMeasureMeta = OptionMeta & {
  layer: string;
  hint: string;
  derived: boolean;
  /** The base measures this one stands for. Empty for `net`, which stands for none. */
  parts: string[];
};

export type JvOptions = {
  posting_bases: OptionMeta[];
  split_modes: OptionMeta[];
  detail_levels: OptionMeta[];
  sign_conventions: OptionMeta[];
  voucher_date_rules: OptionMeta[];
  rounding_modes: OptionMeta[];
  net_pay_sources: OptionMeta[];
  export_formats: OptionMeta[];
  group_by: OptionMeta[];
  sides: OptionMeta[];
  measures: JvMeasureMeta[];
};

export type JvRule = {
  id?: string;
  sequence?: number;
  label: string;
  account_code: string;
  account_name: string | null;
  side: "debit" | "credit";
  measures: string[];
  filters: Record<string, string[]>;
  cost_center_from: string | null;
  active: boolean;
  note?: string | null;
};

export type JvTemplate = {
  id: string;
  name: string;
  note: string | null;
  state: "draft" | "approved";
  is_current: boolean;
  posting_basis: string;
  split_mode: string;
  group_by: string | null;
  detail_level: string;
  sign_convention: string;
  net_pay_source: string;
  voucher_date_rule: string;
  voucher_type: string;
  narration_template: string;
  export_format: string;
  balance_tolerance: number;
  rounding_mode: string;
  rounding_account: string | null;
  approved_by: string | null;
  approved_at: string | null;
  created_at: string | null;
  rule_count: number;
  rules?: JvRule[];
};

export type JvPreset = {
  key: string;
  name: string;
  note: string;
  posting_basis: string;
  split_mode: string;
  detail_level: string;
  group_by: string | null;
  sign_convention: string;
  net_pay_source: string;
  rules: JvRule[];
  rule_count: number;
};

export type VoucherLine = {
  account_code: string;
  account_name: string;
  label: string;
  side: "debit" | "credit";
  amount: number;
  debit: number;
  credit: number;
  cost_center: string | null;
  measures: string[];
  employee_id: string | null;
  employee_name: string | null;
};

export type Voucher = {
  number: string;
  date: string;
  type: string;
  scope: string | null;
  narration: string;
  lines: VoucherLine[];
  total_debit: number;
  total_credit: number;
  difference: number;
  balanced: boolean;
};

export type JvDocument = {
  period: string;
  period_label: string;
  options: Record<string, unknown>;
  employee_count: number;
  vouchers: Voucher[];
  total_debit: number;
  total_credit: number;
  difference: number;
  balanced: boolean;
  warnings: string[];
  cost_totals: Record<string, number>;
  template?: JvTemplate;
};

export type JvReconciliation = JvDocument & {
  counts: ExceptionCounts;
  exceptions: ReconException[];
  summary: {
    vouchers: number;
    lines: number;
    total_debit: number;
    total_credit: number;
    difference: number;
    payroll_cost: number;
    employee_count: number;
    reconciled: boolean;
  };
};

// ---------------------------------------------------------------------------
// Runs and the month as a whole
// ---------------------------------------------------------------------------
export type ReconRun = {
  id: string;
  period: string;
  period_label: string;
  kind: string;
  state: "open" | "closed";
  summary: Record<string, unknown>;
  bank_file_id: string | null;
  jv_template_id: string | null;
  exception_count: number;
  created_by: string | null;
  created_at: string | null;
  closed_by: string | null;
  closed_at: string | null;
  exceptions?: ReconException[];
};

export type Overview = {
  period: string | null;
  period_label?: string;
  ready?: boolean;
  message?: string;
  periods: { key: string; label: string }[];
  register?: { employees: number; net_due: number; stated_net_available: boolean };
  bank?: {
    files: BankFileMeta[];
    paid_total: number;
    ready: boolean;
    message: string | null;
  };
  jv?: { template: JvTemplate | null; ready: boolean; message: string | null };
  runs?: ReconRun[];
  reconciled?: boolean;
};

// ---------------------------------------------------------------------------
// Calls
// ---------------------------------------------------------------------------
const base = "/api/reconciliation";

function get<T>(path: string): Promise<T> {
  return apiFetch(`${base}${path}`).then((r) => parseEnvelopeResponse<T>(r));
}

export const fetchOverview = (period?: string) =>
  get<Overview>(`/overview${period ? `?period=${encodeURIComponent(period)}` : ""}`);

export const fetchBankFields = () => get<BankFieldCatalogue>("/bank/fields");

export const fetchBankPresets = () =>
  get<{ presets: BankPreset[]; caveat: string }>("/bank/presets");

export const fetchProfiles = () => get<{ profiles: BankProfile[] }>("/bank/profiles");

export const fetchBankFiles = (period?: string) =>
  get<{ files: BankFileMeta[] }>(`/bank/files${period ? `?period=${encodeURIComponent(period)}` : ""}`);

export const fetchBankFile = (id: string) =>
  get<BankFileMeta & { rows: BankFileRow[]; truncated: boolean }>(`/bank/files/${id}`);

export const fetchBankReconciliation = (fileId: string) =>
  get<BankReconciliation>(`/bank/reconcile?file_id=${fileId}`);

export const fetchJvOptions = () => get<JvOptions>("/jv/options");

export const fetchJvPresets = () =>
  get<{ presets: JvPreset[]; caveat: string }>("/jv/presets");

export const fetchTemplates = () => get<{ templates: JvTemplate[] }>("/jv/templates");

export const fetchTemplate = (id: string) => get<JvTemplate>(`/jv/templates/${id}`);

export const fetchJvReconciliation = (period: string, templateId?: string) =>
  get<JvReconciliation>(
    `/jv/reconcile?period=${encodeURIComponent(period)}${templateId ? `&template_id=${templateId}` : ""}`,
  );

export const fetchRuns = (period?: string) =>
  get<{ runs: ReconRun[] }>(`/runs${period ? `?period=${encodeURIComponent(period)}` : ""}`);

export const fetchRun = (id: string) => get<ReconRun>(`/runs/${id}`);

export const fetchExceptionKinds = () => get<{ kinds: ExceptionKind[] }>("/exceptions");

export function saveProfile(body: unknown, id?: string) {
  return apiFetch(`${base}/bank/profiles${id ? `/${id}` : ""}`, {
    method: id ? "PUT" : "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  }).then((r) => parseEnvelopeResponse<BankProfile>(r));
}

export function deleteProfile(id: string) {
  return apiFetch(`${base}/bank/profiles/${id}`, { method: "DELETE" }).then((r) =>
    parseEnvelopeResponse<{ deleted: boolean }>(r),
  );
}

export function testProfile(file: File, profile: unknown) {
  const form = new FormData();
  form.append("file", file);
  form.append("profile_json", JSON.stringify(profile));
  return apiFetch(`${base}/bank/profiles/test`, { method: "POST", body: form }).then((r) =>
    parseEnvelopeResponse<ProfileTest>(r),
  );
}

export function suggestMapping(file: File, delimiter = ",") {
  const form = new FormData();
  form.append("file", file);
  form.append("delimiter", delimiter);
  return apiFetch(`${base}/bank/profiles/suggest`, { method: "POST", body: form }).then((r) =>
    parseEnvelopeResponse<MappingSuggestion>(r),
  );
}

export function uploadBankFile(file: File, period: string, profileId: string, note?: string) {
  const form = new FormData();
  form.append("file", file);
  form.append("period", period);
  form.append("profile_id", profileId);
  if (note) form.append("note", note);
  return apiFetch(`${base}/bank/files`, { method: "POST", body: form }).then((r) =>
    parseEnvelopeResponse<BankFileMeta>(r),
  );
}

export function deleteBankFile(id: string) {
  return apiFetch(`${base}/bank/files/${id}`, { method: "DELETE" }).then((r) =>
    parseEnvelopeResponse<{ deleted: boolean }>(r),
  );
}

export function saveTemplate(body: unknown, id?: string) {
  return apiFetch(`${base}/jv/templates${id ? `/${id}` : ""}`, {
    method: id ? "PUT" : "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  }).then((r) => parseEnvelopeResponse<JvTemplate>(r));
}

export function approveTemplate(id: string) {
  return apiFetch(`${base}/jv/templates/${id}/approve`, { method: "POST" }).then((r) =>
    parseEnvelopeResponse<JvTemplate>(r),
  );
}

export function deleteTemplate(id: string) {
  return apiFetch(`${base}/jv/templates/${id}`, { method: "DELETE" }).then((r) =>
    parseEnvelopeResponse<{ deleted: boolean }>(r),
  );
}

export function runReconciliation(body: {
  period_month: string;
  kind: "bank" | "jv";
  bank_file_id?: string;
  jv_template_id?: string;
}) {
  return apiFetch(`${base}/runs`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  }).then((r) => parseEnvelopeResponse<ReconRun>(r));
}

export function closeRun(id: string) {
  return apiFetch(`${base}/runs/${id}/close`, { method: "POST" }).then((r) =>
    parseEnvelopeResponse<ReconRun>(r),
  );
}

// ---------------------------------------------------------------------------
// Presentation
// ---------------------------------------------------------------------------
/**
 * Severity, as a tone.
 *
 * High is danger because money may have moved wrongly. Low is deliberately
 * neutral rather than a softer red: a page where everything is red is a page
 * nobody reads the red on.
 */
export const SEVERITY_TONE: Record<Severity, string> = {
  high: "bg-danger-500/10 text-danger-700 ring-1 ring-danger-500/30 dark:text-danger-300",
  medium: "bg-warning-500/10 text-warning-800 ring-1 ring-warning-500/30 dark:text-warning-200",
  low: "bg-ink-500/10 text-ink-600 ring-1 ring-ink-500/20 dark:text-ink-300",
};

export const SEVERITY_LABEL: Record<Severity, string> = {
  high: "High",
  medium: "Medium",
  low: "Low",
};

/** The current month, as the period pickers here write it. */
export function currentPeriod(): string {
  const now = new Date();
  return `${now.getFullYear()}-${String(now.getMonth() + 1).padStart(2, "0")}`;
}

export function periodKey(iso: string): string {
  return iso.slice(0, 7);
}
